src/codebase.md

```diff
@@ -202,26 +202,42 @@ If arenas still sound like a nasty workaround, consider that you might choose an
 * All of the objects in the arena will be freed at the end of the arena's lifetime, instead of during their manipulation, which can give very low latency for some use-cases. [Bumpalo](https://docs.rs/bumpalo/3.6.1/bumpalo/) formalizes this.
 * The rest of your program might have real Rust references into the arena. You can give the arena a named lifetime (`\`arena` for example), making the provenance of those references very clear.
 
-## I'm having a miserable time making my data structure due to the borrow checker.
+## I'm having a miserable time making my data structure. Should I use unsafe?
 
-If you're writing a data structure, you've probably made poor life
-choices in the first place. You should be using somebody else's tried-and-tested
-data structure. [petgraph](https://docs.rs/petgraph) and
-[slotmap](https://docs.rs/slotmap) are great examples. Use someone else's crate
-by default, and resort to writing your own only if you exhaust that option.
+Low-level data structures are _hard_ in Rust. Arguably, Rust merely makes plain
+all the lifetime and ownership issues which you already had in other languages, but
+the compiler is brutal about it, and you're going to have a bad day.
+
+Even something as simple as a
+[doubly-linked list is notoriously hard](https://rust-unofficial.github.io/too-many-lists/).
+As that (wonderful) book makes clear, you are often faced with a choice:
```

Comment:
Without clicking through the link, this reads a bit awkwardly as it's not clear from the text alone that the thing the link points to is a book. Maybe:

Even something as simple as a doubly-linked list is notoriously hard; so much so that there is a [book that teaches Rust based solely on linked lists](https://rust-unofficial.github.io/too-many-lists/). As that (wonderful) book makes clear, you are often faced with a choice:

Reply:
Done, thanks